crates/ra_fmt2/src/dsl.rs

```diff
@@ -0,0 +1,198 @@
+use crate::pattern::{Pattern, PatternSet};
+use crate::trav_util::{next_non_whitespace_sibling, prev_non_whitespace_sibling};
+use ra_syntax::{SyntaxElement, SyntaxKind::*};
+
+/// `SpacingRule` describes whitespace requirements between `SyntaxElement` Note
+/// that it doesn't handle indentation (first whitespace on a line), there's
+/// `IndentRule` for that!
+#[derive(Debug, Clone)]
+pub(crate) struct SpacingRule {
+    /// An element to which this spacing rule applies
+    pub(crate) pattern: Pattern,
+    /// How much space to add/remove at the start or end of the element.
+    pub(crate) space: Space,
+}
+
+/// Make `SpacingRule` usable with `PatternSet`
+impl AsRef<Pattern> for SpacingRule {
+    fn as_ref(&self) -> &Pattern {
+        &self.pattern
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct Space {
+    /// How much space to add.
+    pub(crate) value: SpaceValue,
+    /// Should the space be added before, after or around the element?
+    pub(crate) loc: SpaceLoc,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum SpaceValue {
+    /// Single whitespace char, like ` `
+    Single,
+    /// Single whitespace char, like ` `, but preserve existing line break.
+    SingleOptionalNewline,
+    /// A single newline (`\n`) char
+    Newline,
+    /// No whitespace at all.
+    None,
+    /// No space, but preserve existing line break.
+    NoneOptionalNewline,
+    /// If the parent element fits into a single line, a single space.
+    /// Otherwise, at least one newline.
+    /// Existing newlines are preserved.
+    SingleOrNewline,
+    /// If the parent element fits into a single line, no space.
+    /// Otherwise, at least one newline.
+    /// Existing newlines are preserved.
+    NoneOrNewline,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum SpaceLoc {
+    /// Before the element.
+    Before,
+    /// After the element.
+    After,
+    /// On the both sides of the element.
+    Around,
+}
+
+/// A builder to conveniently specify a set of `SpacingRule`s
+#[derive(Debug, Default)]
+pub(crate) struct SpacingDsl {
+    pub(crate) rules: Vec<SpacingRule>,
+    #[cfg(test)]
+    pub(crate) tests: Vec<(&'static str, &'static str)>,
```

Comment:
Please document this field and the `test` method. It is not immediately clear what it is used for.